apiserver/facades/client/application/deployrepository.go

```diff
@@ -80,3 +140,467 @@ func validateDeployFromRepositoryArgs(_ params.DeployFromRepositoryArg) []error
+		origin:          resolvedOrigin,
+		placement:       arg.Placement,
+		storage:         stateStorageConstraints(arg.Storage),
+	}
+	if !resolvedCharm.Meta().Subordinate {
+		dt.constraints = arg.Cons
+	}
+	deployRepoLogger.Tracef("validateDeployFromRepositoryArgs returning: %s", pretty.Sprint(dt))
+	return dt, errs
+}
+
+type caasDeployFromRepositoryValidator struct {
+	validator *deployFromRepositoryValidator
+}
+
+func (v caasDeployFromRepositoryValidator) ValidateArg(arg params.DeployFromRepositoryArg) (deployTemplate, []error) {
+	// TODO: NumUnits
+	// TODO: Storage
+	dt, errs := v.validator.validate(arg)
+
+	if corecharm.IsKubernetes(dt.charm) && charm.MetaFormat(dt.charm) == charm.FormatV1 {
+		deployRepoLogger.Debugf("DEPRECATED: %q is a podspec charm, which will be removed in a future release", arg.CharmName)
+	}
+	return dt, errs
+}
+
+type iaasDeployFromRepositoryValidator struct {
+	validator *deployFromRepositoryValidator
+}
+
+func (v *iaasDeployFromRepositoryValidator) ValidateArg(arg params.DeployFromRepositoryArg) (deployTemplate, []error) {
+	// TODO: NumUnits
+	// TODO: Storage
+	return v.validator.validate(arg)
+}
+
+func (v *deployFromRepositoryValidator) createOrigin(arg params.DeployFromRepositoryArg) (*charm.URL, corecharm.Origin, bool, error) {
+	path, err := charm.EnsureSchema(arg.CharmName, charm.CharmHub)
+	if err != nil {
+		return nil, corecharm.Origin{}, false, err
+	}
+	curl, err := charm.ParseURL(path)
+	if err != nil {
+		return nil, corecharm.Origin{}, false, err
+	}
+	if arg.Revision != nil {
+		curl = curl.WithRevision(*arg.Revision)
+	}
+	if !charm.CharmHub.Matches(curl.Schema) {
+		return nil, corecharm.Origin{}, false, errors.Errorf("unknown schema for charm URL %q", curl.String())
+	}
+	channelStr := corecharm.DefaultChannelString
+	if arg.Channel != nil && *arg.Channel != "" {
+		channelStr = *arg.Channel
+	}
+	channel, err := charm.ParseChannelNormalize(channelStr)
+	if err != nil {
+		return nil, corecharm.Origin{}, false, err
+	}
+
+	plat, usedModelDefaultBase, err := v.deducePlatform(arg)
+	if err != nil {
+		return nil, corecharm.Origin{}, false, err
+	}
+
+	origin := corecharm.Origin{
+		Channel:  &channel,
+		Platform: plat,
+		Revision: arg.Revision,
+		Source:   corecharm.CharmHub,
+	}
+	return curl, origin, usedModelDefaultBase, nil
+}
+
+// deducePlatform returns a platform for initial resolveCharm call.
+// At minimum, it must contain an architecture.
+// Platform is determined by the args: architecture constraint and
+// provided base.
+// - Check placement to determine known machine platform. If diffs from
+// other provided data return error.
+// - If no base provided, use model default base.
+// - If no model default base, will be determined later.
+// - If no architecture provided, use model default. Fallback
+// to DefaultArchitecture.
+func (v *deployFromRepositoryValidator) deducePlatform(arg params.DeployFromRepositoryArg) (corecharm.Platform, bool, error) {
```

Comment:
Looks like the `utils` package have some logic about this as well, it might be useful to move these and consolidate the logic over there.

Reply:
There are differences with how we're deducing a platform here and in the client code. A big one is checking the existing machines if placement specified as such. Otherwise I'd agree. That has been done in a few places already.